### PR TITLE
KAFKA-14299: Initialize tasks in state updater

### PR DESCRIPTION
**Pull request by lucasbru**
The state updater code path put tasks into an "initialization queue", with created, but not initialized tasks. These are later, during the event-loop, initialized and added to the state updater.

This leads to problems, and I do not see a reason why this is necessary - we can initialize the task immediately. It can cause losing track of tasks - in particular in the state updater code path it is possible to create tasks twice, if we do not go once around `runLoop` to initialize the task. This can lead to `IllegalStateExceptions`. 

This PR fixes this behavior by removing the initialization queue and instead initializing the tasks lazily inside the state updater thread.

For testing, I adapted the existing unit tests. I will add an integration test once this code path is stable

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by lucasbru** [comment]
@cadonna Could you have a look?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 1]
```diff
@@ -395,8 +395,10 @@ private void createNewTasks(final Map<TaskId, Set<TopicPartition>> activeTasksTo
             tasks.addActiveTasks(newActiveTasks);
             tasks.addStandbyTasks(newStandbyTask);
         } else {
-            tasks.addPendingTaskToInit(newActiveTasks);
-            tasks.addPendingTaskToInit(newStandbyTask);
+            final Map<TaskId, RuntimeException> taskInitExceptions = new LinkedHashMap<>();
+            Stream.concat(newActiveTasks.stream(), newStandbyTask.stream())
+                    .forEach(t -> addTaskToStateUpdater(t, taskInitExceptions));
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: When I started working on the state updater I had the same thought, but then I learned that task initialization might take some time due to the need of opening RocksDB instances. `handleAssignment()` is called in the  callback `ConsumerPartitionAssignor#onAssignment()` which is invoked when a group member receives its assignment from the leader. Having such time consuming operations on the rebalance path is not ideal. I do not remember exactly the reason, but I think with cooperative rebalancing the client might miss the immediate next rebalance which would lead again to more rebalances.
@ableegoldman Please keep me honest here.

**Comment by lucasbru** [comment]
If that is the case, especially if it takes a long time to init rocksdb,
maybe it would be the br see to initialize the state in the state updater
instead?

Bruno Cadonna ***@***.***> schrieb am Di., 1. Nov. 2022,
18:56:

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 1 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: If that is the case, especially if it takes a long time to init rocksdb, maybe it would be easier to initialize the state in the state updater instead?

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Yeah, we considered also to initialise the tasks inside the state updater, but we decided to do it outside so that we do not need to change the state of a task inside the state updater. It seemed cleaner to us to keep lifecycle management away from the state updater. The state updater should only update the states.
What are the advantages of moving the initialisation into the state updater? I see that we would not need to explicitly handle exceptions during intialisation of tasks since that would happen automatically within the state updater. Is there something else?

Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: If we initialize inside the state updater (lazily using `initializeIfNeeded`, before restoration), we should be able to avoid any extra handling for these tasks. In `handleAssignment`, we can directly add the uninitialized tasks to the state updater.

The advantages would be
- No `pendingTaskToInit` required, so some simplification.
- We don't crash when two rebalances happen in one round of polling (why I opened this PR, see description)
- After a rebalance, all tasks that remained with this instance can immediately continue processing, and we do not have to wait for new tasks to initialize their rocksdb (this is more the cherry on top)

The last option that I see to fix the `IllegalStateException` (besides moving initialization to the poll-phase or to the state updater) is that we can take care to not recreate tasks that are already in `pendingTaskToInit`, but it could get complicated. It can happen that the input partitions change, in which case we need to move the already created task but not initialized task to `pendingTaskToUpdateInputPartitions` - but at the same time remember that the task isn't initialized yet ... not sure exactly how to handle these things correctly, but it looks like it will just add more corner cases to think about.

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: I think I am convinced. Let's give initializing the tasks inside the state updater a try.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java [hunk 2]
```diff
@@ -115,6 +117,7 @@ public void run() {
 
         private void runOnce() throws InterruptedException {
             performActionsOnTasks();
+            initializeTasksIfNeeded();
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: Why not initialize the task during `addTask()`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java [hunk 3]
```diff
@@ -277,6 +294,9 @@ private void addTask(final Task task) {
                     }
                 }
             }
+            if (task.state() == Task.State.CREATED) {
+                tasksToInitialize.offer(task);
+            }
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: I would have expected that you initialize the task before any other action, i.e. as first step in this method. Before this PR, we also initialized the task before we passed it to the state updater. Are there any reasons to do it this way?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 4]
```diff
@@ -1908,6 +1888,7 @@ public void shouldThrowWhenHandlingClosingTasksOnProducerCloseError() {
         expectLastCall();
         activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(taskId00);
         expectLastCall().andThrow(new RuntimeException("KABOOM!"));
+        expect(activeTaskCreator.createTasks(anyObject(), anyObject())).andReturn(mkSet());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: How is this related to the change?

**Review by cadonna — COMMENTED** [review]
@lucasbru Thanks for the update!

Here my feedback!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java (hunk 2 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: I wanted to avoid keeping the `tasksAndActionsLock` for too long. Based on what you said initialization can take some time, so there is some potential for blocking the main thread here. This is more or less just defensive programming and I don't have a very concrete situation where this becomes a problem. One can imagine that if you get two rebalances in a row, the second rebalance will be blocked in `handleAssignment` if the initialization from the first rebalance is still blocking `tasksAndActionsLock`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java (hunk 3 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: Well, only that the order here doesn't matter because I use the queue for the reason I mention above.

In principle, I can restore the original order (to be safe), but just moving all the logic in this method to initializeTasksIfNeeded (except for the addition to the queue of course). wdyt?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 4 above)
Comment (lucasbru) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Sorry, that is a leftover

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java (hunk 2 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: Makes sense!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java (hunk 3 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: With this order, you register the task's changelogs and state manager before you initialize the task. Did you verify whether that causes any side-effects?

I think we also have a concurrency issue here. In case of a stateless task, we add the task to the queue of restored tasks but keep a reference to the task in the state updater for initialization. It could be that the main thread reads the stateless task from the queue of restored tasks, but the state updater thread still needs to intialize the task.

So I think in `addTask()` we can only add the task to `tasksToInitialize` and initialize the task as well as register it to the changelog reader or adding it to the queue of restored tasks in a different method similar to `initializeTasksIfNeeded()` (maybe named differently) that does not hold the lock on the input queue. 

If we do this we can maybe also get rid of the verifications whether the state updater is enabled to avoid registering the changelogs to the changelog reader during intialization as done here:
https://github.com/apache/kafka/blob/a1f3c6d16061566a4f53c72a95e2679b8ee229e0/streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java#L215
and remove the explicit changelog registration in the state updater. Could you check this?

Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: > With this order, you register the task's changelogs and state manager before you initialize the task. Did you verify whether that causes any side-effects?

I don't think it should make a difference, but I restored the order in the current version anyways
 
> I think we also have a concurrency issue here. In case of a stateless task, we add the task to the queue of restored tasks but keep a reference to the task in the state updater for initialization. It could be that the main thread reads the stateless task from the queue of restored tasks, but the state updater thread still needs to intialize the task.
> 
> So I think in `addTask()` we can only add the task to `tasksToInitialize` and initialize the task as well as register it to the changelog reader or adding it to the queue of restored tasks in a different method similar to `initializeTasksIfNeeded()` (maybe named differently) that does not hold the lock on the input queue.

Yes, this is the solution I proposed yesterday, and we agreed to do it in the call today. However, during implementation I realized that there are a lot of complications related to `tasksToInitialize` - in many places we now have to handle the tasks from both `updatingTasks` and `tasksToInitialize`, and it felt like an over-complication to me. 

I propose an alternative solution to the problem: Instead of keeping an additional queue, we simply acquire and release the `tasksAndActions` lock for every item of `tasksAndActions` separately, so that we can initialize the task outside the critical section. So this solution should solve all problems we discussed:

 - Initialization does not block the main thread
 - Stateless tasks are only returned to the main thread after initialization has finished
 - It's rather trivial to handle exceptions during initialization, because we remove tasks from `tasksAndActions` one by one and do not have multiple tasks in-flight.
 - The order of initialization and changelog registration should be the same as before, so we are safe if changing the order would cause any problems.

> If we do this we can maybe also get rid of the verifications whether the state updater is enabled to avoid registering the changelogs to the changelog reader during intialization as done here:
> 
> https://github.com/apache/kafka/blob/a1f3c6d16061566a4f53c72a95e2679b8ee229e0/streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java#L215
> 
> 
> and remove the explicit changelog registration in the state updater. Could you check this?

To me, it seems that it's cleaner to leave this as is. Changelogs are registered and unregistered when tasks are added/removed. Initialization only happens when the task is added _AND_ its state is CREATED. So the two operations have different lifetimes - changelogs are only registered when the task is owned by the state updater, but tasks remain initialized after being returned to the main thread. We could attempt to change this behavior (essentially not unregister changelogs when tasks are removed from the state updater, but I suppose this behavior was added for a reason). As long as we want to keep this behavior, I think it is cleaner to handle both registration and unregistration directly inside StateUpdater.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java [hunk 5]
```diff
@@ -121,21 +121,32 @@ private void runOnce() throws InterruptedException {
...
                     switch (action) {
                         case ADD:
-                            addTask(taskAndAction.getTask());
+                            final Task task = taskAndAction.getTask();
+                            final Task existingTask = updatingTasks.putIfAbsent(task.id(), task);
+                            if (existingTask != null) {
+                                throw new IllegalStateException((existingTask.isActive() ? "Active" : "Standby") + " task " + task.id() + " already exist, " +
+                                    "should not try to add another " + (task.isActive() ? "active" : "standby") + " task with the same id. " + BUG_ERROR_MESSAGE);
+                            }
+                            toInitialize = task;
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: Could you extract this code to a method that returns a boolean that can be assigned to `toInitialize`? I think that makes the code more readable.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java [hunk 6]
```diff
@@ -250,30 +261,31 @@ private void removeUpdatingAndPausedTasks() {
...
+                } catch (final StreamsException streamsException) {
+                    addToExceptionsAndFailedTasksThenRemoveFromUpdatingTasks(new ExceptionAndTasks(Collections.singleton(task), streamsException));
+                    initializationFailed = true;
                 }
-                changelogReader.register(task.changelogPartitions(), task.stateManager());
-                if (task.isActive()) {
-                    log.info("Stateful active task " + task.id() + " was added to the state updater");
-                    changelogReader.enforceRestoreActive();
+            }
+            if (!initializationFailed) {
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: Why do we need this flag? Adding the code in the `if`-branch to the `try`-block after `task.initializeIfNeeded();` should have the same effect without needing a flag.
For readability, I would extract the code that starts with `if (isStateless(task))` to a separate method named `handleInitializedTask()` or `postInitializeTask()` or similar.

**Review by cadonna — COMMENTED** [review]
@lucasbru Thanks for the updates!

I like your proposal!

Here my feedback!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java (hunk 3 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: I like your proposal and I agree with what you said about registering to the changelog reader.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java (hunk 6 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: I introduced the flag to make the try-block smaller. For me, it's easier to reason about this code this way, because I know that the exception can only be thrown inside `initializeIfNeeded` and the task is in a well-defined state in the whole try-block (it's inside `updatingTasks` and no other queue). If I move the code starting with `if (isStateless(task))` into the `try`-block, I have to reason about things like - what if the exception happens somewhere inside the stateless task handling? Is the task maybe already removed from `updatingTasks`? But you are right, since in the current code no `StreamsException` is produced anywhere the in the code starting with `if(isStateless(task))` it does not make a difference at runtime. So let me know if you'd prefer me to change it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java (hunk 5 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: boolean won't quite work because then I don't have a reference to the task in the initialization block. 

I introduced a function for adding the task though.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java (hunk 6 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: I see what you mean. You could just return in the `catch`-block after the exception is added to the exception queue.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java (hunk 5 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: Yeah, sorry! My bad! I meant more or less what you did, actually!

**Review by cadonna — APPROVED** [review]
@lucasbru Thanks!

LGTM!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java (hunk 6 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java]: Good idea, done.

**Comment by lucasbru** [comment]
I noticed we then also shouldn't initialize the task during recycling (to not get races between state updater and main thread). This was flagged during smoke testing

**Comment by cadonna** [comment]
Good catch!

**Comment by lucasbru** [comment]
The failed tests are unrelated. 

```
Build / JDK 11 and Scala 2.13 / testSecondaryRefreshAfterElapsedDelay() – org.apache.kafka.common.security.oauthbearer.internals.secured.RefreshingHttpsJwksTest
5s
Build / JDK 17 and Scala 2.13 / [1] tlsProtocol=TLSv1.2, useInlinePem=false – org.apache.kafka.common.network.SslTransportLayerTest
```

I did see `DefaultStateUpdaterTest` fail before (before merging trunk). But I reran it 100 times locally, and it did not happen, and now it doesn't happen in CI anymore.

**Comment by cadonna** [comment]
Build failures are not related:
```
Build / JDK 11 and Scala 2.13 / org.apache.kafka.common.security.oauthbearer.internals.secured.RefreshingHttpsJwksTest.testSecondaryRefreshAfterElapsedDelay()
Build / JDK 17 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest.[1] tlsProtocol=TLSv1.2, useInlinePem=false
```
